genesyscloud/data_source_genesyscloud_routing_skill_group.go

```diff
@@ -31,52 +27,74 @@ func dataSourceRoutingSkillGroup() *schema.Resource {
 }
 
 func dataSourceRoutingSkillGroupRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var (
+		resp *platformclientv2.APIResponse
+		err  error
+	)
 
 	sdkConfig := m.(*provider.ProviderMeta).ClientConfig
 	routingAPI := platformclientv2.NewRoutingApiWithConfig(sdkConfig)
-
 	name := d.Get("name").(string)
 
-	// Find first non-deleted skill by name. Retry in case new skill is not yet indexed by search
-	return util.WithRetries(ctx, 15*time.Second, func() *retry.RetryError {
-		for pageNum := 1; ; pageNum++ {
-			const pageSize = 100
+	retryErr := util.WithRetries(ctx, 15*time.Second, func() *retry.RetryError {
+		var skillGroups []platformclientv2.Skillgroupdefinition
+		skillGroups, resp, err = getAllSkillGroupsByName(routingAPI, name)
+		if err != nil {
+			return retry.NonRetryableError(err)
+		}
 
-			apiClient := &routingAPI.Configuration.APIClient
-			path := routingAPI.Configuration.BasePath + "/api/v2/routing/skillgroups"
+		for _, skillGroup := range skillGroups {
+			if name == *skillGroup.Name {
+				d.SetId(*skillGroup.Id)
+				return nil
+			}
+		}
 
-			headerParams := buildHeaderParams(routingAPI)
-			response, err := apiClient.CallAPI(path, "GET", nil, headerParams, nil, nil, "", nil)
+		return retry.RetryableError(fmt.Errorf("failed to find skill group with name '%s'", name))
+	})
 
-			if err != nil {
-				return retry.RetryableError(util.BuildWithRetriesApiDiagnosticError("genesyscloud_routing_skill_group", fmt.Sprintf("error encountered while trying to retrieve routing skills group found with name %s | error: %s", name, err), response))
-			}
+	if retryErr != nil {
+		msg := fmt.Sprintf("failed to read skill group by name %s: %v", name, retryErr)
+		if resp != nil {
+			return util.BuildAPIDiagnosticError(getSkillGroupResourceName(), msg, resp)
+		}
+		return util.BuildDiagnosticError(getSkillGroupResourceName(), msg, fmt.Errorf("%v", retryErr))
+	}
 
-			if err == nil && response.Error != nil {
-				return retry.RetryableError(util.BuildWithRetriesApiDiagnosticError("genesyscloud_routing_skill_group", fmt.Sprintf("error encountered while trying to retrieve routing skills group found with name %s | error:%s", name, err), response))
-			}
-			if err == nil && response.StatusCode == http.StatusNotFound {
-				return retry.RetryableError(util.BuildWithRetriesApiDiagnosticError("genesyscloud_routing_skill_group", fmt.Sprintf("routing skills group not found with name %s", name), response))
-			}
+	return nil
+}
 
-			allSkillGroups := &AllSkillGroups{}
+func getAllSkillGroupsByName(api *platformclientv2.RoutingApi, name string) ([]platformclientv2.Skillgroupdefinition, *platformclientv2.APIResponse, error) {
+	const pageSize = 100
+	var (
+		after          string
+		allSkillGroups []platformclientv2.Skillgroupdefinition
+	)
+
+	for {
```

Comment:
is it not worth looking for pagecount and then looping through ?

some APIs do not properly return the Entities size object

Reply:
This API doesn't return a page count. If there is another page, nextUri will exist in the response, if it is the last page, it won't. That is the only indicator